pre_commit/repository.py

```diff
@@ -64,11 +64,10 @@ def languages(self):
 
     @cached_property
     def additional_dependencies(self):
-        dep_dict = defaultdict(lambda: defaultdict(set))
+        dep_dict = defaultdict(lambda: defaultdict(list))
         for _, hook in self.hooks:
-            dep_dict[hook['language']][hook['language_version']].update(
-                hook.get('additional_dependencies', []),
-            )
+            dep_dict[hook['language']][hook['language_version']] += \
+                hook.get('additional_dependencies', [])
```

Comment:
Well, the current `set` implementation is already accepting the same dependency listed at different versions. We could keep the same behaviour by handling duplicates manually:
```python
for dep in hook.get('additional_dependencies', []):
    if dep not in dep_dict[hook['language']][hook['language_version']]:
        dep_dict[hook['language']][hook['language_version']].append(dep)
```

An other option would be to extract the name and the version for each dependency (needs to be implemented for each language though) and then either:
  - keep only the first version of each package
  - issue a warning to the user
  - raise an exception

and then implement tests accordingly, what would you prefer?